cEP-0033.md

```diff
@@ -0,0 +1,760 @@
+`HTML and Jinja` , `Python and Jinja` , `codeblocks and RST` etc.
+
+coala does not yet support these kind of files. This project would enable coala
+to deal with those situations and allow people to write code analysis similar to
+how they already do it while being applicable to the right locations at the
+right files.The users of coala would not have to concentrate on writing new 
+bears/ analysis routines. This implementation would perfectly work with the 
+existing bears. 
+
+There can be several ways to approach this. In this cEP, we will be implementing
+a abstract way which will support arbitrary combination of languages.
+
+## A higher level view of the implementation
+
+A source file containing multiple programming languages is loaded up by coala. 
+The original nested file is broken/split into `n` temporary files, where n is
+the number of languages present in the original file. Each temporary file only
+contains the snippets belonging to one language. 
+
+These temporary files would then be passed to their respective language bears 
+( _chosen by the user_) where the static analysis routines would run. The
+temporary files after being linted would then be assembled. 
+
+The above process creates the illusion of the original file is being linted, 
+but in reality we divide the files into different parts and lint them one by 
+one.
+
+The figure below, would help give a clear picture
+
+![higher_level_view](https://github.com/Naveenaidu/gsoc-18-materials/blob/master/photos/high_level_view_handle_nested.jpg)
+
+## Nested Language Section (nl_section)
+
+The original file is segregated into small pieces/units where each piece/unit
+purely contains only `one language`. These quantum pieces of the the nested 
+original file is termed as `nl_section`.
+
+A `nl_section` can further be defined as a group of lines in the source code that
+purely belongs to one particular language. 
+
+All the `nl_sections` that belong to one particular language are grouped to form
+a valid file of that programming language.
+
+The following figure would make the things clear
+
+![nl_sections](https://github.com/Naveenaidu/gsoc-18-materials/blob/master/photos/file_section.jpg)
+
+
+In the above figure, we have a file which contains both HTML and PHP code. This
+original file can be broken up into 4 different sections.
+
+Each nl_section will contain the following information:
+
+1. The Programming language of the lines
+2. Index of the section
+3. Starting Line number in the original file
+4. Ending Line number in the original file
+5. Starting line number in the linted file
+6. Ending line number in the linted file
+
+# Prototype of nl_section
+
+```python
+class NlSection():
+
+    def __init__(self, orig_start, orig_end, language, index, linted_start=None, 
+                 linted_end =None, temp_file=None):
+        """
+        Creates a new NlSection.
+
+        :param orig_start:   The start line number of the nl_section
```

Comment:
I was also thinking of adding a column field for each line.
```python
:param orig_start_column: The column of the starting line
```

So that we can cover the cases where we have the languages on the same line. What are your opinions about this?

Reply:
Yeah a start column would be good :+1: (thinking about indentation and inline statements)